From f0d4e43dab5e86286a94d0669ab3c6a2d74ab907 Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Wed, 18 Feb 2015 09:55:40 -0800
Subject: [PATCH] iio: inv_mpu6050: do dmi_check_system() once for ACPI systems

DMI system matching needs to be only done once. This fixes
an issue where the kernel panics when unbinding/binding
driver to device. This was caused by the callbacks being
defined as __init, and they got freed after module init.
So when a binding is being done, dmi_check_system() is
calling functions that no longer exist in memory,
causing panic.

Change-Id: I7971c03f1e962a1b24ba7009dab54fd812dd9c2b
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 drivers/iio/imu/inv_mpu6050/inv_mpu_acpi.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/iio/imu/inv_mpu6050/inv_mpu_acpi.c b/drivers/iio/imu/inv_mpu6050/inv_mpu_acpi.c
index 0fd54e3..35d0d95 100644
--- a/drivers/iio/imu/inv_mpu6050/inv_mpu_acpi.c
+++ b/drivers/iio/imu/inv_mpu6050/inv_mpu_acpi.c
@@ -20,11 +20,12 @@
 
 enum inv_mpu_product_name {
 	INV_MPU_NOT_MATCHED,
+	INV_MPU_OTHERS,
 	INV_MPU_ASUS_T100TA,
 	INV_MPU_MALATA,
 };
 
-static enum inv_mpu_product_name matched_product_name;
+static enum inv_mpu_product_name matched_product_name = INV_MPU_NOT_MATCHED;
 
 static int __init asus_t100_matched(const struct dmi_system_id *d)
 {
@@ -174,7 +175,9 @@ int inv_mpu_acpi_create_mux_client(struct inv_mpu6050_state *st)
 		adev = ACPI_COMPANION(&st->client->dev);
 		memset(&info, 0, sizeof(info));
 
-		dmi_check_system(inv_mpu_dev_list);
+		if (matched_product_name == INV_MPU_NOT_MATCHED)
+			dmi_check_system(inv_mpu_dev_list);
+
 		switch (matched_product_name) {
 		/* Malata has the same BIOS implementation with T100TA */
 		case INV_MPU_MALATA:
@@ -184,6 +187,8 @@ int inv_mpu_acpi_create_mux_client(struct inv_mpu6050_state *st)
 			break;
 		/* Add more matched product processing here */
 		default:
+			/* no matched platform */
+			matched_product_name = INV_MPU_OTHERS;
 			break;
 		}
 
-- 
1.9.1

